drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+
+    # Calculate the global low watermark
+    wmark_low = 0
+    for zone in for_each_zone(prog):
+        if has_member(zone, "_watermark"):
+            zone_low_wmark = zone.member_("_watermark")[
+                enum_value_get(prog, "WMARK_LOW")
+            ]
+        else:
+            zone_low_wmark = zone.member_("watermark")[
+                enum_value_get(prog, "WMARK_LOW")
+            ]
+        wmark_low += zone_low_wmark.value_()
+
+    # Estimate reclaimable page cache
+    lru_pages = get_lru_pages(prog)
+    lru_active_file = lru_pages[enum_value_get(prog, "LRU_ACTIVE_FILE")]
+    lru_inactive_file = lru_pages[enum_value_get(prog, "LRU_INACTIVE_FILE")]
+    pagecache = lru_active_file + lru_inactive_file
+    pagecache -= min(pagecache // 2, wmark_low)
+    available_pages += pagecache
+
+    # Determine the number of reclaimable pages
+    reclaimable_pages = 0
+    # Estimate reclaimable slab cache
+    slab_reclaimable = get_global_page_state(prog, "NR_SLAB_RECLAIMABLE")
+    reclaimable_pages += slab_reclaimable
+
+    if "NR_KERNEL_MISC_RECLAIMABLE" in prog:
+        misc = get_global_page_state(prog, "NR_KERNEL_MISC_RECLAIMABLE")
+        reclaimable_pages += misc
+
+    reclaimable_pages -= min(reclaimable_pages // 2, wmark_low)
+    available_pages += reclaimable_pages
+
+    # Estimate reclaimable kernel memory
+    if "NR_INDIRECTLY_RECLAIMABLE_BYTES" in prog:
+        page_shift = prog["PAGE_SHIFT"].value_()
+        indirect_reclaimable_pages = (
+            get_global_page_state(prog, "NR_INDIRECTLY_RECLAIMABLE_BYTES")
+            >> page_shift
+        )
+        available_pages += indirect_reclaimable_pages
+
+    available_pages = max(0, available_pages)
+    return available_pages
+
+
+def get_total_swapcache_pages(prog: Program) -> int:
+    """
+    Get the total number of cached swap pages for all swap types.
+
+    :returns: The number of cached memory back by swap space in pages.
+    """
+    ret = 0
+    nr_swapper_spaces = prog["nr_swapper_spaces"]
+    swapper_spaces = prog["swapper_spaces"]
+
+    for i in range(_MAX_SWAPFILES):
+        nr = nr_swapper_spaces[i]
+        spaces = swapper_spaces[i]
+        if nr == 0 or spaces.value_() == 0x0:
+            continue
+
+        for j in range(nr):
+            ret += spaces[j].member_("nrpages").value_()
+    return ret
+
+
+def get_lru_pages(prog: Program) -> List[int]:
+    """
+    Get the number of pages for all LRU lists, including: INACTIVE_ANON,
+    ACTIVE_ANON, INACTIVE_FILE, ACTIVE_FILE, and LRU_UNEVICTABLE.
+
+    :returns: A list of integer numbers that represents numbers of pages
+    in each LRU list.
+    """
+    try:
+        lru_pages = prog.cache["lru_pages"]
+    except KeyError:
+        num_lru_lists = prog.constant("NR_LRU_LISTS").value_()
+        lru_pages = [0 for _ in range(num_lru_lists)]
+        for lru in range(num_lru_lists):
+            lru_pages[lru] = get_global_page_state(prog, "NR_LRU_BASE", lru)
```

Comment:
I would argue that rather than doing a list comprehension, followed by a loop, you could do this in one comprehension.
```suggestion
        lru_pages = [
            get_global_page_state(prog, "NR_LRU_BASE", lru)
            for lru in range(num_lru_lists)
        ]
```
It's not any shorter but I think it makes a bit more sense on reading.

Reply:
Makes sense. Thanks. Updated.